File: app/test/gamification/badge_engine_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:gym_coach/gamification/badge_engine.dart';

void main() {
  test('first workout unlocks FIRST_WORKOUT badge', () {
    final unlocked = checkUnlockedBadges({
      'totalWorkoutsCompleted': 1,
      'longestStreak': 0,
      'totalDaysAllMealsCompleted': 0,
      'level': 1,
    });
    expect(unlocked, contains('FIRST_WORKOUT'));
  });

  test('7 day streak unlocks ONE_WEEK_STREAK', () {
    final unlocked = checkUnlockedBadges({
      'totalWorkoutsCompleted': 7,
      'longestStreak': 7,
      'totalDaysAllMealsCompleted': 0,
      'level': 1,
    });
    expect(unlocked, contains('ONE_WEEK_STREAK'));
  });

  test('100 day streak unlocks HUNDRED_DAYS', () {
    final unlocked = checkUnlockedBadges({
      'totalWorkoutsCompleted': 100,
      'longestStreak': 100,
      'totalDaysAllMealsCompleted': 0,
      'level': 5,
    });
    expect(unlocked, contains('HUNDRED_DAYS'));
  });

  test('7 consecutive all-meal days unlocks IRON_STOMACH', () {
    final unlocked = checkUnlockedBadges({
      'totalWorkoutsCompleted': 7,
      'longestStreak': 7,
      'totalDaysAllMealsCompleted': 7,
      'level': 1,
    });
    expect(unlocked, contains('IRON_STOMACH'));
  });

  test('level 10 unlocks FITNESS_OWL', () {
    final unlocked = checkUnlockedBadges({
      'totalWorkoutsCompleted': 50,
      'longestStreak': 30,
      'totalDaysAllMealsCompleted': 0,
      'level': 10,
    });
    expect(unlocked, contains('FITNESS_OWL'));
  });

  test('no badges for new user', () {
    final unlocked = checkUnlockedBadges({
      'totalWorkoutsCompleted': 0,
      'longestStreak': 0,
      'totalDaysAllMealsCompleted': 0,
      'level': 1,
    });
    expect(unlocked, isEmpty);
  });
}

File: app/android/app/src/test/java/com/gymcoach/gym_coach/gamification/BadgeEngineTest.kt
package com.gymcoach.gym_coach.gamification

import com.google.common.truth.Truth.assertThat
import org.junit.Test

class BadgeEngineTest {

    @Test
    fun `new user gets no badges`() {
        assertThat(BadgeEngine.checkUnlockedBadges(emptyMap())).isEmpty()
    }

    @Test
    fun `first workout badge unlocks at 1`() {
        val stats = mapOf(
            "totalWorkoutsCompleted" to 1,
            "longestStreak" to 0,
            "totalDaysAllMealsCompleted" to 0,
            "level" to 1
        )
        assertThat(BadgeEngine.checkUnlockedBadges(stats)).containsExactly(BadgeEngine.FIRST_WORKOUT)
    }

    @Test
    fun `one week streak badge unlocks at 7`() {
        val stats = mapOf(
            "totalWorkoutsCompleted" to 0,
            "longestStreak" to 7,
            "totalDaysAllMealsCompleted" to 0,
            "level" to 1
        )
        assertThat(BadgeEngine.checkUnlockedBadges(stats)).contains(BadgeEngine.ONE_WEEK_STREAK)
    }

    @Test
    fun `hundred days badge unlocks at 100`() {
        val stats = mapOf(
            "totalWorkoutsCompleted" to 0,
            "longestStreak" to 100,
            "totalDaysAllMealsCompleted" to 0,
            "level" to 1
        )
        assertThat(BadgeEngine.checkUnlockedBadges(stats)).contains(BadgeEngine.HUNDRED_DAYS)
    }

    @Test
    fun `iron stomach badge unlocks after 7 all-meal days`() {
        val stats = mapOf(
            "totalWorkoutsCompleted" to 0,
            "longestStreak" to 0,
            "totalDaysAllMealsCompleted" to 7,
            "level" to 1
        )
        assertThat(BadgeEngine.checkUnlockedBadges(stats)).contains(BadgeEngine.IRON_STOMACH)
    }

    @Test
    fun `fitness owl badge unlocks at level 10`() {
        val stats = mapOf(
            "totalWorkoutsCompleted" to 0,
            "longestStreak" to 0,
            "totalDaysAllMealsCompleted" to 0,
            "level" to 10
        )
        assertThat(BadgeEngine.checkUnlockedBadges(stats)).contains(BadgeEngine.FITNESS_OWL)
    }
}
